### PR TITLE
Add simple CPU offloading support.

### PR DESCRIPTION
**Pull request by janimo**
Based on the vLLM implementation, applied to llama.py, gemma2.py and qwen2.py

Tested with unquantized Llama 3.1 8B, Gemma 2 9B and Qwen 2.5 7B on a 6GB RTX 4060.

### CODE REVIEW

**Comment by janimo** [comment]
This is the code it is based on https://github.com/vllm-project/vllm/pull/6496

File: test/srt/test_srt_engine.py [hunk 1]
```diff
@@ -160,6 +160,38 @@ def test_7_engine_offline_throughput(self):
...
+        prompt = "Today is a sunny day and I like"
+        model_path = DEFAULT_SMALL_MODEL_NAME_FOR_TEST
+
+        sampling_params = {"temperature": 0, "max_new_tokens": 8}
+
+        engine = sgl.Engine(
+            model_path=model_path,
+            random_seed=42,
+            max_total_tokens=128,
+            log_level="error",
```
Comment (merrymercy) [inline test/srt/test_srt_engine.py]:
```suggestion
```

File: test/srt/test_srt_engine.py [hunk 2]
```diff
@@ -160,6 +160,38 @@ def test_7_engine_offline_throughput(self):
...
+        )
+        out1 = engine.generate(prompt, sampling_params)["text"]
+        engine.shutdown()
+
+        engine = sgl.Engine(
+            model_path=model_path,
+            random_seed=42,
+            max_total_tokens=128,
+            log_level="error",
```
Comment (merrymercy) [inline test/srt/test_srt_engine.py]:
```suggestion
```

**Review by merrymercy — APPROVED** [review]

**Comment by zhyncs** [comment]
@janimo @merrymercy Unit test 3 failed

**Comment by janimo** [comment]
> @janimo @merrymercy Unit test 3 failed

@zhyncs it fails in a different test, before the one included in this PR, it seems to be an unrelated OOM present in other CI runs as well.

**Pull request merged**